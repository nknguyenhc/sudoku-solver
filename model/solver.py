from copy import deepcopy

class InvalidSudokuException(Exception):
    def __init__(self, i, j):
        super().__init__(f"No value available at ({i}, {j})")

class Sudoku:
    """A class to store the board, list of assignments and reduced domains"""
    def __init__(self, board, domains=None):
        """Assume that board and domains are already copied and is not used elsewhere
        The board is a 9x9 board, where filled numbers are ints, and unfilled are None.
        Domains is a dictionary of keys of coordinates and values are the corresponding domain.
        Eg: {(0, 1): {1, 2, 3}}
        """
        self.board = board
        if domains:
            self.domains = domains
        else:
            self.domains = Sudoku.generate_default_domains(board)
    
    def generate_default_domains(board):
        """Generate the default domain for unfilled cells for the given board,
        when domain is not given
        """
        domains = dict()
        for i in range(9):
            for j in range(9):
                if board[i][j] is not None:
                    continue
                domain = {1, 2, 3, 4, 5, 6, 7, 8, 9}
                domain = Sudoku.reduce_domain(domain, board, i, j)
                domains[(i, j)] = domain
        return domains
    
    def reduce_domain(domain, board, i, j):
        domain = Sudoku.reduce_domain_in_row(domain, board, i)
        domain = Sudoku.reduce_domain_in_column(domain, board, j)
        domain = Sudoku.reduce_domain_in_group(domain, board, i, j)
        if len(domain) == 0:
            raise InvalidSudokuException(i, j)
        return domain
    
    def reduce_domain_in_row(domain, board, i):
        for j in range(9):
            if board[i][j] in domain:
                domain.remove(board[i][j])
        return domain
    
    def reduce_domain_in_column(domain, board, j):
        for i in range(9):
            if board[i][j] in domain:
                domain.remove(board[i][j])
        return domain
    
    def reduce_domain_in_group(domain, board, row, col):
        start_i = 3 * (row // 3)
        start_j = 3 * (col // 3)
        for i in range(start_i, start_i + 3):
            for j in range(start_j, start_j + 3):
                if board[i][j] in domain:
                    domain.remove(board[i][j])
        return domain
    
    def assign(self, cell, value):
        """Assign the cell to the value,
        returning a new instance of Sudoku and leave this instance intact
        """
        board = deepcopy(self.board)
        domains = deepcopy(self.domains)
        domains.pop(cell, None)
        board[cell[0]][cell[1]] = value
        sudoku = Sudoku(board, domains)
        sudoku.propagate_constraints()
        return sudoku
    
    def propagate_constraints(self):
        """Update the domains based on the board"""
        for cell, domain in self.domains.items():
            updated_domain = Sudoku.reduce_domain(
                domain, self.board, cell[0], cell[1])
            self.domains[cell] = updated_domain
    
    def __eq__(self, other):
        if not isinstance(other, Sudoku):
            return False
        return self.board == other.board
    
    def is_terminal(self):
        """Check if the board is completed
        This means that the domains dict is empty.
        """
        return len(self.domains) == 0
    
    def minimum_domain_cell(self):
        min_cell = None
        domain_size = 10
        for cell, domain in self.domains.items():
            curr_size = len(domain)
            if curr_size == 1:
                return cell
            if curr_size < domain_size:
                domain_size = curr_size
                min_cell = cell
        return min_cell


class SudokuSolver:
    """An agent to solve Sudoku, as CSP"""
    def __init__(self):
        pass
    
    def solve(self, sudoku):
        if sudoku.is_terminal():
            return sudoku
        cell = sudoku.minimum_domain_cell()
        for value in sudoku.domains[cell]:
            try:
                new_sudoku = sudoku.assign(cell, value)
            except InvalidSudokuException:
                continue
            result = self.solve(new_sudoku)
            if result is not None:
                return result
        return None
