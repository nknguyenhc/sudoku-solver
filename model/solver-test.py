from solver import InvalidSudokuException, Sudoku, SudokuSolver

def get_empty_sudoku():
    board = list(list(None for i in range(9)) for j in range(9))
    return Sudoku(board)

def read_file(filename):
    """Returns the board from the file"""
    def read_char(char):
        if char == '_':
            return None
        else:
            return int(char)
        
    def readline(f):
        line = f.readline().strip()
        cells = [read_char(char) for char in line.split(' ')]
        return cells

    with open(f"tests/{filename}", 'r') as f:
        return [readline(f) for i in range(9)]

def test_empty_board():
    sudoku = get_empty_sudoku()
    for i in range(9):
        for j in range(9):
            assert sudoku.domains[(i, j)] == {1, 2, 3, 4, 5, 6, 7, 8, 9}

def test_reduced_domain():
    sudoku = get_empty_sudoku()
    sudoku = (
        sudoku.assign((0, 0), 1)
        .assign((0, 1), 2)
        .assign((1, 0), 3)
        .assign((1, 1), 5)
        .assign((2, 1), 7)
        .assign((0, 5), 3)
        .assign((7, 2), 8)
    )
    assert sudoku.domains[(0, 2)] == {4, 6, 9}

def test_not_reduced_domain():
    sudoku = get_empty_sudoku()
    sudoku = (
        sudoku.assign((0, 0), 1)
        .assign((0, 1), 3)
        .assign((1, 3), 5)
        .assign((2, 3), 6)
        .assign((5, 1), 9)
    )
    assert sudoku.domains[(0, 2)] == {2, 4, 5, 6, 7, 8, 9}

def exception_raised():
    sudoku = get_empty_sudoku()
    sudoku = (
        sudoku.assign((0, 0), 1)
        .assign((0, 1), 2)
        .assign((1, 0), 3)
        .assign((1, 1), 4)
        .assign((1, 2), 5)
        .assign((2, 1), 6)
        .assign((2, 2), 7)
        .assign((0, 5), 8)
    )
    try:
        sudoku.assign((0, 8), 9)
        assert False
    except Exception as e:
        assert isinstance(e, InvalidSudokuException)
    try:
        sudoku.assign((2, 0), 9)
        assert False
    except Exception as e:
        assert isinstance(e, InvalidSudokuException)
    try:
        sudoku.assign((7, 2), 9)
        assert False
    except Exception as e:
        assert isinstance(e, InvalidSudokuException)

def test_minimum_domain_cell():
    sudoku = get_empty_sudoku()
    sudoku = (
        sudoku.assign((1, 0), 1)
        .assign((1, 1), 3)
        .assign((6, 2), 4)
        .assign((2, 7), 8)
    )
    assert sudoku.minimum_domain_cell() == (2, 2)

def test_solve_case(casename):
    print(f"Testing case {casename}")
    in_board = read_file(f"{casename}.in")
    result = SudokuSolver().solve(Sudoku(in_board))
    out_board = read_file(f"{casename}.out")
    assert Sudoku(out_board) == result

def main():
    test_empty_board()
    test_reduced_domain()
    test_not_reduced_domain()
    test_minimum_domain_cell()
    test_solve_case("easy")
    test_solve_case("medium")
    test_solve_case("hard")
    test_solve_case("expert")
    test_solve_case("master")

if __name__ == '__main__':
    main()
