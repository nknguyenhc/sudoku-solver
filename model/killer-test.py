from killer import KillerConstraint, KillerSudoku
from solver import SudokuSolver

def test_constraints_0():
    constraint = KillerConstraint({}, 7)
    assert constraint.available_values == set()

def test_constraints_1():
    constraint = KillerConstraint({(0, 0)}, 6)
    assert constraint.available_values == {6}
    constraint = KillerConstraint({(0, 0)}, 10)
    assert constraint.available_values == set()

def test_constraints_2():
    constraint = KillerConstraint({(0, 0), (0, 1)}, 2)
    assert constraint.available_values == set()
    constraint = KillerConstraint({(0, 0), (0, 1)}, 6)
    assert constraint.available_values == {1, 2, 4, 5}
    constraint = KillerConstraint({(0, 0), (0, 1)}, 11)
    assert constraint.available_values == {2, 3, 4, 5, 6, 7, 8, 9}
    constraint = KillerConstraint({(0, 0), (0, 1)}, 17)
    assert constraint.available_values == {8, 9}
    constraint = KillerConstraint({(0, 0), (0, 1)}, 18)
    assert constraint.available_values == set()

def test_constraints_3():
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2)}, 5)
    assert constraint.available_values == set()
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2)}, 6)
    assert constraint.available_values == {1, 2, 3}
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2)}, 15)
    assert constraint.available_values == {1, 2, 3, 4, 5, 6, 7, 8, 9}
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2)}, 20)
    assert constraint.available_values == {3, 4, 5, 6, 7, 8, 9}
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2)}, 23)
    assert constraint.available_values == {6, 8, 9}
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2)}, 25)
    assert constraint.available_values == set()

def test_constraints_more():
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)}, 45)
    assert constraint.available_values == {1, 2, 3, 4, 5, 6, 7, 8, 9}
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)}, 46)
    assert constraint.available_values == set()
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)}, 44)
    assert constraint.available_values == set()
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1)}, 35)
    assert constraint.available_values == set()
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1)}, 37)
    assert constraint.available_values == {1, 2, 3, 4, 5, 6, 7, 9}

def test_contains():
    constraint = KillerConstraint({(0, 0), (2, 2)}, 3)
    assert constraint.contains((0, 0))
    assert not constraint.contains((2, 0))

def test_remove_cell():
    constraint = KillerConstraint({(0, 0), (2, 2)}, 5)
    constraint = constraint.remove_cell((0, 0), 2)
    assert constraint.coords == {(2, 2)}
    assert constraint.value == 3
    assert constraint.available_values == {3}
    constraint = KillerConstraint({(0, 0), (0, 1), (0, 2)}, 10)
    constraint = constraint.remove_cell((0, 0), 2)
    assert constraint.coords == {(0, 1), (0, 2)}
    assert constraint.value == 8
    assert constraint.available_values == {1, 2, 3, 5, 6, 7}

def read_constraint(string):
    """Read the constraint from the string representation
    The string representation is as follows:
    "x0 y0 x1 y1 ... xn yn v", where (xi, yi) are coordinates of the cells,
    and v is the sum of the cells"""
    nums = string.split(" ")
    coords = set()
    for i in range(len(nums) // 2):
        x = int(nums[2 * i])
        y = int(nums[2 * i + 1])
        coords.add((x, y))
    v = int(nums[-1])
    return KillerConstraint(coords, v)

def read_file(filename):
    """Returns an instance of KillerSudoku from the file"""
    def read_char(char):
        if char == '_':
            return None
        else:
            return int(char)
    
    def readline(f):
        line = f.readline().strip()
        cells = [read_char(char) for char in line.split(' ')]
        return cells
    
    with open(f"tests/{filename}", 'r') as f:
        board = [readline(f) for i in range(9)]
        line = f.readline()
        constraints = []
        while line:
            line = line.strip()
            constraint = read_constraint(line)
            constraints.append(constraint)
            line = f.readline()
        return KillerSudoku(board, constraints)

def test_solve_case(casename):
    print(f"Testing case {casename}")
    in_sudoku = read_file(f"killer.{casename}.in")
    result = SudokuSolver().solve(in_sudoku)
    out_sudoku = read_file(f"killer.{casename}.out")
    assert out_sudoku == result

def main():
    test_constraints_0()
    test_constraints_1()
    test_constraints_2()
    test_constraints_3()
    test_constraints_more()
    test_contains()
    test_remove_cell()
    test_solve_case("easy")
    test_solve_case("medium")
    test_solve_case("hard")
    test_solve_case("expert")

if __name__ == '__main__':
    main()
