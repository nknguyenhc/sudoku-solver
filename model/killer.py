from copy import deepcopy
from model.solver import Sudoku, InvalidSudokuException

class KillerConstraint():
    def __init__(self, coords, value):
        """Instantiate a new constraint
        Coords is a set of coordinates, each is a tuple of 2 elements representing the cell coordinates.
        Value is an integer represent sum of the cells."""
        self.coords = coords
        self.value = value
        self.available_values = self.calc_available_values()
    
    def calc_available_values(self):
        """Calculate the possible values that can be used to fulfil the sum
        Each number must be unique."""
        count = len(self.coords)
        if count == 0:
            return set()
        elif count == 1:
            return set([self.value]) if (self.value >= 1 and self.value <= 9) else set()

        values = [i for i in range(count - 1)]
        i = 0
        domain = set()
        while True:
            if i == 0:
                if values[i] == 9 - count + 1:
                    break
                else:
                    values[i] += 1
                    i += 1
            elif i < count - 1:
                if values[i] == 9 - count + i + 1:
                    values[i] = 0
                    i -= 1
                elif values[i] == 0:
                    values[i] = values[i - 1] + 1
                    i += 1
                else:
                    values[i] += 1
                    i += 1
            else:
                last_value = self.value - sum(values)
                if last_value > values[i - 1] and last_value <= 9:
                    new_domain = set(values)
                    new_domain.add(last_value)
                    domain = domain.union(new_domain)
                i -= 1
        return domain
    
    def contains(self, coord):
        return coord in self.coords
    
    def remove_cell(self, cell, value):
        coords = deepcopy(self.coords)
        coords.remove(cell)
        if len(coords) == 0:
            return None
        return KillerConstraint(coords, self.value - value)
    
    def __hash__(self):
        return hash(self.coords)
    
    def __eq__(self, other):
        if not isinstance(other, KillerConstraint):
            return False
        return self.coords == other.coords
    
    def __str__(self):
        return f"Coords: {self.coords}, Sum: {self.value}, Possible values: {self.available_values}"


class KillerSudoku(Sudoku):
    """A class to store the board, list of assignments, reduced domains and the constraints"""
    def __init__(self, board, constraints, domains=None, is_updated=False):
        """Assume that the board, domains and constraints are already copied and is not used elsewhere
        The constraints is a list of instances of KillerConstraint class above"""
        super().__init__(board, domains=domains)
        self.constraints = constraints
        if not is_updated:
            self.domains = KillerSudoku.update_domains_from_constraints(self.domains, self.constraints)
            self.domains = KillerSudoku.reduce_domains_in_constraint_groups(self.domains, self.board, self.constraints)
    
    def update_domains_from_constraints(domains, constraints):
        """Update the domains of the cells based on the constraints"""
        for constraint in constraints:
            for cell in constraint.coords:
                if cell not in domains:
                    continue
                domains[cell] = domains[cell].intersection(constraint.available_values)
                if len(domains[cell]) == 0:
                    raise InvalidSudokuException(cell[0], cell[1])
        return domains
    
    def reduce_domains_in_constraint_groups(domains, board, constraints):
        for constraint in constraints:
            for value_cell in constraint.coords:
                if value_cell in domains:
                    continue
                for cell in constraint.coords:
                    if cell not in domains:
                        continue
                    domains[cell].discard(board[value_cell[0]][value_cell[1]])
                    if len(domains[cell]) == 0:
                        raise InvalidSudokuException(cell[0], cell[1])
        return domains
    
    def reduce_domain_in_constraint_group(domains, constraints, value):
        for constraint in constraints:
            for cell in constraint.coords:
                if cell not in domains:
                    continue
                domains[cell].discard(value)
                if len(domains[cell]) == 0:
                    raise InvalidSudokuException(cell[0], cell[1])
        return domains
    
    def assign(self, cell, value):
        board = deepcopy(self.board)
        board[cell[0]][cell[1]] = value

        domains = deepcopy(self.domains)
        domains.pop(cell, None)

        constraints = list(filter(
            lambda constraint: not constraint.contains(cell),
            self.constraints))
        removed_constraints = filter(
            lambda constraint: constraint.contains(cell),
            self.constraints)
        new_constraints = list(filter(
            lambda constraint: constraint is not None,
            map(lambda constraint: constraint.remove_cell(cell, value),
                removed_constraints)))
        
        domains = KillerSudoku.reduce_domain_in_constraint_group(domains, new_constraints, value)
        domains = KillerSudoku.update_domains_from_constraints(domains, new_constraints)
        constraints = constraints + new_constraints
        
        sudoku = KillerSudoku(board, constraints, domains=domains, is_updated=True)
        sudoku.propagate_constraints()
        return sudoku
    
    def __eq__(self, other):
        if not isinstance(other, KillerSudoku):
            return False
        return self.board == other.board
    
    def __str__(self):
        return f"Board:{self.board}\nDomains:{self.domains}\n"
