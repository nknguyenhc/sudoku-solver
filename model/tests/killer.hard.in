_ _ 2 _ _ _ _ _ _
_ _ _ _ _ 3 _ _ _
_ _ _ _ _ _ _ _ 1
_ _ _ _ 8 _ _ _ _
1 _ _ _ _ _ _ _ _
_ _ _ 2 _ _ _ _ 4
_ _ _ _ _ _ _ 2 _
_ 7 _ _ _ _ _ _ _
_ _ _ _ _ _ _ 8 _
0 0 0 1 1 0 21
0 2 1 2 2 2 3 2 22
0 3 1 3 1 4 2 4 21
0 4 0 5 1 5 9
0 6 0 7 0 8 1 7 2 7 28
1 1 2 0 2 1 3 1 18
1 6 2 6 10
1 8 2 8 3 8 9
2 3 9
2 5 3 4 3 5 19
3 0 4 0 5 0 10
3 3 6
3 6 4 6 6
3 7 3
4 1 5 1 10
4 2 4 3 13
4 4 4 5 5 5 16
4 7 4 8 17
5 2 5 3 6 3 10
5 4 6 4 7
5 6 5 7 13
5 8 6 8 11
6 0 9
6 1 7 1 8 1 14
6 2 7 2 8 2 14
6 5 6 6 12
6 7 7 6 7 7 12
7 0 8 0 8
7 3 7 4 7 5 9
7 8 8 7 8 8 16
8 3 8 4 10
8 5 8 6 13