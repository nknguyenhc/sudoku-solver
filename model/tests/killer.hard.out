7 6 2 8 5 1 3 4 9
8 1 9 4 7 3 2 5 6
3 5 4 9 2 6 8 7 1
4 9 7 6 8 5 1 3 2
1 2 6 7 3 4 5 9 8
5 8 3 2 1 9 7 6 4
9 3 1 5 6 8 4 2 7
6 7 8 3 4 2 9 1 5
2 4 5 1 9 7 6 8 3